ennead/utils/markdown.py

```diff
@@ -0,0 +1,54 @@
+from typing import Any
+
+import redis
+from markdown import Markdown
+from markdown.extensions import Extension as MarkdownExtension
+
+
+class DisallowHTML(MarkdownExtension):
+    """Simple extension for Python-Markdown that disallows HTML"""
+
+    def extendMarkdown(self, md):
+        md.preprocessors.deregister('html_block')
+        md.inlinePatterns.deregister('html')
+
+
+class DictCache(dict):
+    """Dict with .set() method auto-encoding values"""
+
+    def set(self, key: str, value: str, **_: Any) -> None:
+        """Set value at key to value.encode('utf-8')"""
+
+        self[key] = value.encode('utf-8')
+
+
+class CachedMarkdown:
+    """Cached markdown renderer"""
+
+    def __init__(self):
+        self.engine = Markdown(extensions=['mdx_math', 'fenced_code', DisallowHTML()])
+        try:
+            self.cache_engine = redis.Redis()
+            self.cache_engine.ping()
+        except redis.exceptions.ConnectionError:
+            self.cache_engine = DictCache()
+
+    def render(self, markdown: str) -> str:
+        """Convert Markdown to HTML, or get from cache if already converted"""
+
+        cache_key = f'md:{markdown}'
+        html = self.cache_engine.get(cache_key)
```

Comment:
Если вдруг redis подключится, а затем упадет (ну например, в процессе обновления пакетов на сервере, мало ли вещей админ может сделать, когда на одном сервере много сервисов болтается - и ennead не главный, за которым следят), хорошо бы либо переподключить кэш, либо хотя бы не упасть.

```suggestion
        try:
            html = self.cache_engine.get(cache_key)
        except:
            html = None
```

Reply:
Подумаю над этим